### PR TITLE
Decimal cleanup (#2637)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #2637

# Rationale for this change
 

Follow on to #2857 that takes the opportunity to perform some further cleanups

# What changes are included in this PR?

Makes some APIs infallible, and removes the no longer needed `Decimal` abstraction

# Are there any user-facing changes?

No, #2857 hasn't been released yet

FYI @alamb it would be good for some version of this to make the next release if possible, to avoid API churn

### CODE REVIEW

File: .github/workflows/arrow.yml [hunk 1]
```diff
@@ -186,6 +186,14 @@ jobs:
       - name: Setup Clippy
         run: |
           rustup component add clippy
-      - name: Run clippy
+      - name: Clippy arrow-buffer with all features
```
Comment (tustvold) [inline .github/workflows/arrow.yml:189]: This was a consequence of #2594 and was preventing linting of the split apart crates

File: arrow-array/src/array/primitive_array.rs [hunk 2]
```diff
@@ -917,67 +916,62 @@ impl<T: DecimalType + ArrowPrimitiveType> PrimitiveArray<T> {
...
-        let s = self.scale()?;
-        Ok(Decimal::<T>::new(p, s, &T::to_native(self.value(row))).to_string())
+    /// Returns [`Self::value`] formatted as a string
+    pub fn value_as_string(&self, row: usize) -> String {
+        T::format_decimal(self.value(row), self.precision(), self.scale())
     }
 
-    pub fn precision(&self) -> Result<u8, ArrowError> {
+    /// Returns the decimal precision of this array
+    pub fn precision(&self) -> u8 {
```
Comment (tustvold) [inline arrow-array/src/array/primitive_array.rs:925]: These methods can be made infallible as the error case is actually unreachable, as of #2835 we consistently verify the data type

File: arrow-array/src/array/primitive_array.rs [hunk 3]
```diff
@@ -1564,8 +1558,8 @@ mod tests {
             .build()
             .unwrap();
         let decimal_array = Decimal128Array::from(array_data);
-        assert_eq!(8_887_000_000_i128, decimal_array.value(0).into());
-        assert_eq!(-8_887_000_000_i128, decimal_array.value(1).into());
+        assert_eq!(8_887_000_000_i128, decimal_array.value(0));
```
Comment (tustvold) [inline arrow-array/src/array/primitive_array.rs:1561]: These `Into::into` conversions are redundant and were making clippy unhappy

File: arrow-array/src/types.rs [hunk 4]
```diff
@@ -569,6 +565,26 @@ impl ArrowPrimitiveType for Decimal256Type {
     const DATA_TYPE: DataType = <Self as DecimalType>::DEFAULT_TYPE;
 }
 
+fn format_decimal_str(value_str: &str, precision: usize, scale: usize) -> String {
```
Comment (tustvold) [inline arrow-array/src/types.rs:568]: This is based on the logic from `Decimal::to_string`, adding this allows removing the now unnecessary `Decimal` construct.

**Review by viirya — COMMENTED** [review]
Thanks @tustvold for the cleanups. I will review this in next few days.

**Review by viirya — APPROVED** [review]

**Pull request merged**